File: correct-crypto/cpp/signature.cpp
// Correct Ed25519 signatures (libsodium crypto_sign).
#include <sodium.h>
#include <string>
#include <vector>

bool sign_and_verify(const std::string &message) {
    unsigned char pk[crypto_sign_PUBLICKEYBYTES];
    unsigned char sk[crypto_sign_SECRETKEYBYTES];
    crypto_sign_keypair(pk, sk);

    std::vector<unsigned char> signature(crypto_sign_BYTES);
    unsigned long long sig_len = 0;
    crypto_sign_detached(signature.data(), &sig_len,
                         (const unsigned char *)message.data(), message.size(), sk);
    return crypto_sign_verify_detached(signature.data(),
                                       (const unsigned char *)message.data(),
                                       message.size(), pk) == 0;
}

File: correct-crypto/rust/signature.rs
// Correct Ed25519 signatures (ed25519-dalek crate).
use ed25519_dalek::{Signature, Signer, SigningKey, Verifier};
use rand::rngs::OsRng;

pub fn sign_and_verify(message: &[u8]) -> bool {
    let mut csprng = OsRng;
    let signing_key = SigningKey::generate(&mut csprng);
    let signature: Signature = signing_key.sign(message);
    signing_key.verifying_key().verify(message, &signature).is_ok()
}
